lib/crypto/src/bigint.rs

```diff
@@ -0,0 +1,277 @@
+    /// let one = U64::from(1u64);
+    /// assert_eq!(one.num_bits(), 1);
+    /// let max = U64::from(u64::MAX);
+    /// assert_eq!(max.num_bits(), 64);
+    /// let u32_max = U64::from(u32::MAX as u64);
+    /// assert_eq!(u32_max.num_bits(), 32);
+    /// ```
+    fn num_bits(&self) -> usize;
+
+    /// Compute the `i`-th bit of `self`.
+    /// # Example
+    ///
+    /// ```
+    /// use openzeppelin_crypto::bigint::{BigInteger, crypto_bigint::U64};
+    ///
+    /// let mut one = U64::from(1u64);
+    /// assert!(one.get_bit(0));
+    /// assert!(!one.get_bit(1));
+    /// ```
+    fn get_bit(&self, i: usize) -> bool;
+}
+
+impl<const N: usize> BigInteger for Uint<N> {
+    const NUM_LIMBS: usize = N;
+
+    fn is_odd(&self) -> bool {
+        <Uint<N> as Integer>::is_odd(self).into()
+    }
+
+    fn is_even(&self) -> bool {
+        <Uint<N> as Integer>::is_even(self).into()
+    }
+
+    fn is_zero(&self) -> bool {
+        <Uint<N> as Zero>::is_zero(self).into()
+    }
+
+    fn num_bits(&self) -> usize {
+        self.bits()
+    }
+
+    fn get_bit(&self, i: usize) -> bool {
+        self.bit(i).into()
+    }
+}
+
+impl<const N: usize> BitIteratorBE for Uint<N> {
+    fn bit_be_iter(&self) -> impl Iterator<Item = bool> {
+        self.as_words().iter().rev().flat_map(Word::bit_be_iter)
+    }
+}
+
+/// Parse a number from a string in a given radix.
+///
+/// I.e., convert string encoded integer `s` to base-`radix` number.
+#[must_use]
+pub const fn from_str_radix<const LIMBS: usize>(
+    s: &str,
+    radix: u32,
+) -> Uint<LIMBS> {
+    let bytes = s.as_bytes();
+
+    // The lowest order number is at the end of the string.
+    // Begin parsing from the last index of the string.
+    let mut index = bytes.len() - 1;
+
+    let mut uint = Uint::from_u32(0);
+    let mut order = Uint::from_u32(1);
+    let uint_radix = Uint::from_u32(radix);
+
+    loop {
+        // Try to parse a digit from utf-8 byte
+        let ch = parse_utf8_byte(bytes[index]);
+        let digit = match ch.to_digit(radix) {
+            None => {
+                panic!("invalid digit");
+            }
+            Some(digit) => Uint::from_u32(digit),
+        };
+
+        // Add a digit multiplied by order.
+        uint = add(&uint, &mul(&digit, &order));
+
+        // If we reached the beginning of the string, return the number.
+        if index == 0 {
+            return uint;
+        }
+
+        // Increase the order of magnitude.
+        order = mul(&uint_radix, &order);
+
+        // Move to the next digit.
+        index -= 1;
+    }
+}
+
+/// Multiply two numbers and panic on overflow.
+#[must_use]
+pub const fn mul<const LIMBS: usize>(
+    a: &Uint<LIMBS>,
+    b: &Uint<LIMBS>,
+) -> Uint<LIMBS> {
+    let (low, high) = a.mul_wide(b);
+    assert!(high.bits() == 0, "overflow on multiplication");
+    low
+}
+
+/// Add two numbers and panic on overflow.
+#[must_use]
+pub const fn add<const LIMBS: usize>(
```

Comment:
I do not like this attitude; I believe you should just implement traits like here:
https://doc.rust-lang.org/std/primitive.u64.html#trait-implementations

Reply:
Like [Add](https://doc.rust-lang.org/std/ops/trait.Add.html) trait.

Reply:
And require these traits in `BigInteger`

Reply:
The whole point of this function to have a `const` evaluation.
Const trait functions are not yet supported in rust, and here is the workaround.